# Coding Assignment 2: {insert pun-y name here}

We know we didn't give you a ton of turn around time between the first coding assignment and this coding assignment, but we hope that you'll enjoy doing this assignment, especially if you didn't really like the first one. 

Coding Assignment 2 will be due on **Friday, March 22 at 11:59 PM**. Note: This is the day before we leave for Spring Break. 

As it was for Coding Assignment 1, you may work with **one** other student on this assignment. Please fill in the cell below with the appropriate values, if applicable, for each variable.

In [ ]:
NAME = ""
COLLABORATOR_NAME = ""

We will import the libraries/packages that you may need while doing this assignment. You **SHOULD NOT** change the below cell in any way. 

@Christopher: Make sure to change the metadata of the below cell!

In [1]:
from datascience import *
import numpy as np

Whereas the last assignment was very game theoretical, this assignment will have a very real and tangible outcome, which should give you a better sense of how we might apply basic data tools to confirm or reject your suspicions. Given that this assignment is entirely based on a real situation, we want to remind the class that we, as facilitators, are **NOT** taking a political stance by developing this assignment nor are we telling you that a political stance is better for any particular reason. We are **only** illustrating the idea that we can empirically confirm or reject suspicions that we come across. 

## Introduction

On November 6th, 2018, Mark Harris won the U.S. House of Representatives seat from North Carolina District 9. Upon completing the tally counts, Mark Harris led by **905** (@Christopher: double check this value) out of **130,000** (@Christopher: get the exact value)

We've already done some of the initial steps for you, which can often be the hardest part to any data analytics project. We've provided you with a question (is there evidence of election tampering in Bladen County, North Carolina) and datasets. We've already done a significant portion of the cleaning for you since it can often be highly technical and incredibly frustrating, but we left a portion of this for you to develop an intuition for data cleaning.

In [2]:
election_results = Table.read_table("nc_election_results_11062018.csv")
election_results

County    | Election Date | Precinct | Contest Group ID | Contest Type | Contest Name                                | Choice              | Choice Party | Vote For | Election Day | One Stop | Absentee by Mail | Provisional | Total Votes | Real Precinct

ALAMANCE | 11/06/2018    | 064      | 1228             | S            | NC COURT OF APPEALS JUDGE SEAT 3            | Michael Monaco, Sr. | LIB          | 1        | 59           | 65       | 2                | 1           | 127         | Y

ALAMANCE | 11/06/2018    | 03N      | 1228             | S            | NC COURT OF APPEALS JUDGE SEAT 3            | Michael Monaco, Sr. | LIB          | 1        | 59           | 38       | 1                | 0           | 98          | Y

ALAMANCE | 11/06/2018    | 03S      | 1228             | S            | NC COURT OF APPEALS JUDGE SEAT 3            | Michael Monaco, Sr. | LIB          | 1        | 106          | 108      | 0                | 3           | 217         | Y

ALAMANCE | 11/06/2018    | 04       | 1228             | S            | NC COURT OF APPEALS JUDGE SEAT 3            | Michael Monaco, Sr. | LIB          | 1        | 67           | 38       | 1                | 0           | 106         | Y

ALAMANCE | 11/06/2018    | 035      | 1228             | S            | NC COURT OF APPEALS JUDGE SEAT 3            | Michael Monaco, Sr. | LIB          | 1        | 73           | 54       | 1                | 1           | 129         | Y

ALAMANCE | 11/06/2018    | 07       | 1228             | S            | NC COURT OF APPEALS JUDGE SEAT 3            | Michael Monaco, Sr. | LIB          | 1        | 43           | 33       | 2                | 0           | 78          | Y

ALAMANCE | 11/06/2018    | 09S      | 1228             | S            | NC COURT OF APPEALS JUDGE SEAT 3            | Michael Monaco, Sr. | LIB          | 1        | 83           | 59       | 2                | 0           | 144         | Y

ALAMANCE | 11/06/2018    | 12S      | 1228             | S            | NC COURT OF APPEALS JUDGE SEAT 3            | Michael Monaco, Sr. | LIB          | 1        | 15           | 5        | 0                | 0           | 20          | Y

ALAMANCE | 11/06/2018    | 03N2     | 1273             | S            | NC DISTRICT COURT JUDGE DISTRICT 15A SEAT 1 | Brad Allen          | DEM          | 1        | 217          | 81       | 18               | 0           | 316         | Y

ALAMANCE | 11/06/2018    | 03W      | 1273             | S            | NC DISTRICT COURT JUDGE DISTRICT 15A SEAT 1 | Brad Allen          | DEM          | 1        | 459          | 538      | 20               | 2           | 1019        | Y
... (183544 rows omitted)

One thing you should notice is that all the values in the first column ("County") are the same. Another thing you should notice is that there are 183544 rows omitted. 

**1.1**: How many rows are you in the `election_results` table? How many columns are in the `election_results` table?

In [4]:
print("There are " + str(election_results.num_rows) + " rows.")
print("There are " + str(election_results.num_columns) + " columns.")

There are 183554 rows.
There are 15 columns.


As you can tell, there's a lot of data here. Some of it is pretty useful, but there are 183,554 rows so let's filter out only what's necessary. The dataset here represents all of the contests in the November 2018 elections in North Carolina. We only care about US Congressional District 09, which is the election Mark Harris was involved in. 

**1.2**: Create a new table that only contains data regarding the US House of Representatives District 09 election.

In [5]:
district9 = election_results.where("Contest Name", are.equal_to("US HOUSE OF REPRESENTATIVES DISTRICT 09"))
district9.show(10)

Our dataset is significantly smaller, which makes it easier to work with so that might be able to answer our questions about whether Mark Harris engaged in election fraud. Not all of these columns are useful. **BLAH BLAH BLAH**.  

**1.3**: Drop "Contest Group ID", "Contest Type", and "Vote For" from the `district9` table.

In [7]:
district9 = district9.drop("Contest Group ID", "Contest Type" ,"Vote For")
district9

County  | Election Date | Precinct         | Contest Name                            | Choice       | Choice Party | Election Day | One Stop | Absentee by Mail | Provisional | Total Votes | Real Precinct

BLADEN | 11/06/2018    | P45              | US HOUSE OF REPRESENTATIVES DISTRICT 09 | Dan McCready | DEM          | 60           | 0        | 0                | 0           | 60          | Y

BLADEN | 11/06/2018    | P202             | US HOUSE OF REPRESENTATIVES DISTRICT 09 | Dan McCready | DEM          | 210          | 0        | 0                | 0           | 210         | Y

BLADEN | 11/06/2018    | P60              | US HOUSE OF REPRESENTATIVES DISTRICT 09 | Dan McCready | DEM          | 289          | 0        | 0                | 0           | 289         | Y

BLADEN | 11/06/2018    | P35              | US HOUSE OF REPRESENTATIVES DISTRICT 09 | Dan McCready | DEM          | 89           | 0        | 0                | 0           | 89          | Y

BLADEN | 11/06/2018    | P15              | US HOUSE OF REPRESENTATIVES DISTRICT 09 | Dan McCready | DEM          | 224          | 0        | 0                | 0           | 224         | Y

BLADEN | 11/06/2018    | P201             | US HOUSE OF REPRESENTATIVES DISTRICT 09 | Dan McCready | DEM          | 104          | 0        | 0                | 0           | 104         | Y

BLADEN | 11/06/2018    | PROVISIONAL      | US HOUSE OF REPRESENTATIVES DISTRICT 09 | Dan McCready | DEM          | 0            | 0        | 0                | 5           | 5           | N

BLADEN | 11/06/2018    | ONE STOP LIBRARY | US HOUSE OF REPRESENTATIVES DISTRICT 09 | Dan McCready | DEM          | 0            | 1906     | 0                | 0           | 1906        | N

BLADEN | 11/06/2018    | P40              | US HOUSE OF REPRESENTATIVES DISTRICT 09 | Dan McCready | DEM          | 98           | 0        | 0                | 0           | 98          | Y

BLADEN | 11/06/2018    | P10              | US HOUSE OF REPRESENTATIVES DISTRICT 09 | Dan McCready | DEM          | 103          | 0        | 0                | 0           | 103         | Y
... (728 rows omitted)

@Christopher: there might be more to this?!

## Bladen County

## Other Counties

## Absentee Ballot Data

In [9]:
bladen_absentees = Table.read_table("bladen.csv")
bladen_absentees.show(10)